File: android/app/src/main/java/com/testlogon/android/feature/analysis/MarketStats.kt
package com.testlogon.android.feature.analysis

import kotlin.math.ln
import kotlin.math.sqrt

/**
 * Pure, framework-free market-statistics math for the Analysis workbench. No Android/Compose imports
 * so it can be unit-tested on the JVM (see MarketStatsTest). Every helper takes plain price/return
 * series (chronological, oldest-first) and returns render-ready primitives. Edge cases (empty / single
 * bar / zero baseline) return null or an identity value rather than throwing.
 */
object MarketStats {

    /** ~trading periods per year used to annualize a per-bar volatility (approx calendar days). */
    const val ANNUALIZATION_PERIODS = 365.0

    /** Simple period return (last-first)/first as a fraction. Null when <2 points or a zero baseline. */
    fun periodReturn(closes: List<Double>): Double? {
        if (closes.size < 2) return null
        val first = closes.first()
        if (first == 0.0) return null
        return (closes.last() - first) / first
    }

    /** Alias for [periodReturn] as a percent (x100). */
    fun cumulativeReturnPct(closes: List<Double>): Double? = periodReturn(closes)?.let { it * 100.0 }

    /**
     * Per-step log returns ln(p_i / p_prev), aligned to the gaps between consecutive closes (so a
     * series of N closes yields N-1 returns). Non-positive prices are skipped (log undefined).
     */
    fun logReturns(closes: List<Double>): List<Double> {
        if (closes.size < 2) return emptyList()
        val out = ArrayList<Double>(closes.size - 1)
        for (i in 1 until closes.size) {
            val prev = closes[i - 1]
            val cur = closes[i]
            if (prev > 0.0 && cur > 0.0) out.add(ln(cur / prev))
        }
        return out
    }

    /** Sample standard deviation of a series. Null when fewer than two samples. */
    fun stdev(values: List<Double>): Double? {
        if (values.size < 2) return null
        val mean = values.average()
        val variance = values.sumOf { (it - mean) * (it - mean) } / (values.size - 1)
        return sqrt(variance)
    }

    /**
     * Annualized volatility: sample stdev of log-returns scaled by sqrt([periodsPerYear]). Null when
     * there are not enough returns to form a stdev. Returned as a fraction (0.25 == 25% vol).
     */
    fun annualizedVolatility(
        closes: List<Double>,
        periodsPerYear: Double = ANNUALIZATION_PERIODS,
    ): Double? {
        val sd = stdev(logReturns(closes)) ?: return null
        return sd * sqrt(periodsPerYear)
    }

    /**
     * Maximum drawdown as a POSITIVE fraction: the largest peak-to-trough decline over the series.
     * 0.0 for a monotonically non-decreasing series; null for empty input. A 100 to 50 dip returns 0.5.
     */
    fun maxDrawdown(closes: List<Double>): Double? {
        if (closes.isEmpty()) return null
        var peak = closes.first()
        var maxDd = 0.0
        for (c in closes) {
            if (c > peak) peak = c
            if (peak > 0.0) {
                val dd = (peak - c) / peak
                if (dd > maxDd) maxDd = dd
            }
        }
        return maxDd
    }

    /** Highest close in the series, or null when empty. */
    fun high(closes: List<Double>): Double? = closes.maxOrNull()

    /** Lowest close in the series, or null when empty. */
    fun low(closes: List<Double>): Double? = closes.minOrNull()

    /** Average of a series, or null when empty. */
    fun average(values: List<Double>): Double? = if (values.isEmpty()) null else values.average()

    /** Sum of a series (0.0 when empty). */
    fun total(values: List<Double>): Double = values.sum()

    /**
     * Pearson correlation of two aligned series over their common leading length. Null when fewer
     * than two overlapping points or when either side has zero variance (undefined correlation).
     * Result is clamped to the range minus-one..one to absorb floating-point overshoot.
     */
    fun correlation(a: List<Double>, b: List<Double>): Double? {
        val n = minOf(a.size, b.size)
        if (n < 2) return null
        val xs = a.subList(0, n)
        val ys = b.subList(0, n)
        val mx = xs.average()
        val my = ys.average()
        var sxy = 0.0
        var sxx = 0.0
        var syy = 0.0
        for (i in 0 until n) {
            val dx = xs[i] - mx
            val dy = ys[i] - my
            sxy += dx * dy
            sxx += dx * dx
            syy += dy * dy
        }
        if (sxx == 0.0 || syy == 0.0) return null
        val r = sxy / sqrt(sxx * syy)
        return r.coerceIn(-1.0, 1.0)
    }

    /**
     * Normalize a close series to a base of 100 at its first point, for a multi-symbol overlay. Empty
     * or zero-baseline input returns an empty list (nothing meaningful to overlay).
     */
    fun normalizeToBase(closes: List<Double>, base: Double = 100.0): List<Double> {
        val first = closes.firstOrNull() ?: return emptyList()
        if (first == 0.0) return emptyList()
        return closes.map { it / first * base }
    }

    /**
     * A simple fast/slow SMA-cross backtest over [closes]. Position is LONG when fast SMA > slow SMA,
     * else FLAT; entries/exits act on the NEXT bar (no look-ahead). Returns a [BacktestResult] with
     * the equity curve (starts at 1.0), total return fraction, win rate, and trade count.
     */
    fun backtestMaCross(closes: List<Double>, fast: Int, slow: Int): BacktestResult {
        if (fast <= 0 || slow <= 0 || fast >= slow || closes.size < slow + 1) {
            return BacktestResult(equityCurve = listOf(1.0), totalReturn = 0.0, winRate = 0.0, trades = 0)
        }
        val fastSma = sma(closes, fast)
        val slowSma = sma(closes, slow)

        val equity = ArrayList<Double>(closes.size)
        var cash = 1.0
        equity.add(cash)

        var inPosition = false
        var entryPrice = 0.0
        var trades = 0
        var wins = 0

        // Signal at bar i (using SMAs up to i) is acted on the return from i to i+1.
        for (i in 0 until closes.size - 1) {
            val f = fastSma[i]
            val s = slowSma[i]
            val wantLong = f != null && s != null && f > s

            if (wantLong && !inPosition) {
                inPosition = true
                entryPrice = closes[i]
                trades++
            } else if (!wantLong && inPosition) {
                inPosition = false
                if (closes[i] > entryPrice) wins++
            }

            // Apply this bar-to-next return only while holding.
            if (inPosition && closes[i] != 0.0) {
                cash *= closes[i + 1] / closes[i]
            }
            equity.add(cash)
        }
        // Close any open position at the final bar for win/trade accounting.
        if (inPosition) {
            if (closes.last() > entryPrice) wins++
        }

        val total = cash - 1.0
        val winRate = if (trades == 0) 0.0 else wins.toDouble() / trades
        return BacktestResult(equityCurve = equity, totalReturn = total, winRate = winRate, trades = trades)
    }

    /** Simple moving average aligned 1:1 with [values]; null until [period] samples are available. */
    fun sma(values: List<Double>, period: Int): List<Double?> {
        if (period <= 0) return values.map { null }
        val out = ArrayList<Double?>(values.size)
        var sum = 0.0
        for (i in values.indices) {
            sum += values[i]
            if (i >= period) sum -= values[i - period]
            out.add(if (i >= period - 1) sum / period else null)
        }
        return out
    }
}

/** Result of [MarketStats.backtestMaCross]: the equity curve plus summary stats. */
data class BacktestResult(
    val equityCurve: List<Double>,
    val totalReturn: Double,
    val winRate: Double,
    val trades: Int,
)

File: frontend/src/lib/marketStats.ts
/**
 * Pure, framework-free statistics over a market bar series.
 *
 * A "bar" is the normalized OHLCV shape used by the history endpoint
 * ({@link Bar}); the Analysis workbench maps the exchange `Candle`
 * (integer prices scaled by `price_scaler`, nanosecond timestamps) into this
 * shape before calling in, so everything here is plain floating-point math.
 *
 * Nothing here touches React, the network, or the DOM: it is deterministic and
 * unit-tested in `marketStats.test.ts`.
 */

/** One OHLCV bar. `ts` is epoch-ms; prices/volume are already de-scaled. */
export interface Bar {
  ts: number;
  o: number;
  h: number;
  l: number;
  c: number;
  v: number;
}

/** The number of bars of a given interval that make up one calendar year. */
export function barsPerYear(intervalSec: number): number {
  if (!Number.isFinite(intervalSec) || intervalSec <= 0) return 0;
  const SECONDS_PER_YEAR = 365 * 24 * 60 * 60;
  return SECONDS_PER_YEAR / intervalSec;
}

/** Natural-log returns between consecutive closes (length = bars.length - 1). */
export function logReturns(bars: Bar[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < bars.length; i++) {
    const prev = bars[i - 1]!.c;
    const cur = bars[i]!.c;
    if (prev > 0 && cur > 0) out.push(Math.log(cur / prev));
    else out.push(0);
  }
  return out;
}

/** Simple (arithmetic) returns between consecutive closes. */
export function simpleReturns(bars: Bar[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < bars.length; i++) {
    const prev = bars[i - 1]!.c;
    const cur = bars[i]!.c;
    out.push(prev !== 0 ? cur / prev - 1 : 0);
  }
  return out;
}

/** Arithmetic mean of a numeric array (0 for empty). */
export function mean(xs: number[]): number {
  if (xs.length === 0) return 0;
  let s = 0;
  for (const x of xs) s += x;
  return s / xs.length;
}

/**
 * Sample standard deviation (n-1 denominator). Returns 0 for series with
 * fewer than two observations.
 */
export function stdev(xs: number[]): number {
  const n = xs.length;
  if (n < 2) return 0;
  const m = mean(xs);
  let acc = 0;
  for (const x of xs) {
    const d = x - m;
    acc += d * d;
  }
  return Math.sqrt(acc / (n - 1));
}

/** Simple moving average; entries before `period-1` are null. */
export function sma(vals: number[], period: number): (number | null)[] {
  const out: (number | null)[] = [];
  if (period <= 0) return vals.map(() => null);
  let sum = 0;
  for (let i = 0; i < vals.length; i++) {
    sum += vals[i]!;
    if (i >= period) sum -= vals[i - period]!;
    out.push(i >= period - 1 ? sum / period : null);
  }
  return out;
}

export interface DrawdownResult {
  /** Max drawdown as a negative fraction (e.g. -0.25 = a 25% peak-to-trough fall). */
  maxDrawdown: number;
  /** Index of the peak that precedes the worst trough (-1 if not defined). */
  peakIndex: number;
  /** Index of the worst trough (-1 if not defined). */
  troughIndex: number;
}

/**
 * Maximum drawdown of the close series: the deepest peak-to-trough decline as a
 * negative fraction of the peak. Zero for a non-decreasing series or <2 bars.
 */
export function maxDrawdown(bars: Bar[]): DrawdownResult {
  if (bars.length < 2) return { maxDrawdown: 0, peakIndex: -1, troughIndex: -1 };
  let peak = bars[0]!.c;
  let peakIdx = 0;
  let worst = 0;
  let worstPeakIdx = -1;
  let worstTroughIdx = -1;
  for (let i = 1; i < bars.length; i++) {
    const c = bars[i]!.c;
    if (c > peak) {
      peak = c;
      peakIdx = i;
    } else if (peak > 0) {
      const dd = c / peak - 1; // <= 0
      if (dd < worst) {
        worst = dd;
        worstPeakIdx = peakIdx;
        worstTroughIdx = i;
      }
    }
  }
  return { maxDrawdown: worst, peakIndex: worstPeakIdx, troughIndex: worstTroughIdx };
}

export interface SeriesStats {
  bars: number;
  first: number | null;
  last: number | null;
  hi: number | null;
  lo: number | null;
  /** Total (cumulative) return over the window: last/first - 1. */
  cumulativeReturn: number;
  /** Mean of per-bar simple returns. */
  avgReturn: number;
  /** Sample stdev of per-bar log returns (per-bar, not annualized). */
  volatility: number;
  /** {@link volatility} annualized by sqrt(barsPerYear(interval)). */
  annualizedVolatility: number;
  maxDrawdown: number;
  avgVolume: number;
  totalVolume: number;
}

/** Full stat pack for a bar series at a given interval (seconds). */
export function computeStats(bars: Bar[], intervalSec: number): SeriesStats {
  const n = bars.length;
  if (n === 0) {
    return {
      bars: 0, first: null, last: null, hi: null, lo: null,
      cumulativeReturn: 0, avgReturn: 0, volatility: 0, annualizedVolatility: 0,
      maxDrawdown: 0, avgVolume: 0, totalVolume: 0,
    };
  }
  const first = bars[0]!.c;
  const last = bars[n - 1]!.c;
  let hi = bars[0]!.h;
  let lo = bars[0]!.l;
  let volSum = 0;
  for (const b of bars) {
    if (b.h > hi) hi = b.h;
    if (b.l < lo) lo = b.l;
    volSum += b.v || 0;
  }
  const logRets = logReturns(bars);
  const simpleRets = simpleReturns(bars);
  const perBarVol = stdev(logRets);
  const annFactor = Math.sqrt(barsPerYear(intervalSec));
  return {
    bars: n,
    first,
    last,
    hi,
    lo,
    cumulativeReturn: first !== 0 ? last / first - 1 : 0,
    avgReturn: mean(simpleRets),
    volatility: perBarVol,
    annualizedVolatility: perBarVol * annFactor,
    maxDrawdown: maxDrawdown(bars).maxDrawdown,
    avgVolume: volSum / n,
    totalVolume: volSum,
  };
}

/**
 * Pearson correlation of two equal-length numeric series. Returns 0 when the
 * series differ in length, are shorter than 2, or either has zero variance.
 */
export function correlation(a: number[], b: number[]): number {
  const n = Math.min(a.length, b.length);
  if (n < 2) return 0;
  const ma = mean(a.slice(0, n));
  const mb = mean(b.slice(0, n));
  let cov = 0;
  let va = 0;
  let vb = 0;
  for (let i = 0; i < n; i++) {
    const da = a[i]! - ma;
    const db = b[i]! - mb;
    cov += da * db;
    va += da * da;
    vb += db * db;
  }
  if (va === 0 || vb === 0) return 0;
  return cov / Math.sqrt(va * vb);
}

/**
 * Align two bar series on their shared timestamps and return the paired
 * log-return vectors (one entry per shared consecutive-timestamp pair). Used to
 * correlate two instruments that may not share every bar.
 */
export function alignedLogReturns(a: Bar[], b: Bar[]): { a: number[]; b: number[] } {
  const mapB = new Map<number, number>();
  for (const bar of b) mapB.set(bar.ts, bar.c);
  // Common timestamps in a's order.
  const common: { ts: number; ca: number; cb: number }[] = [];
  for (const bar of a) {
    const cb = mapB.get(bar.ts);
    if (cb != null) common.push({ ts: bar.ts, ca: bar.c, cb });
  }
  const ra: number[] = [];
  const rb: number[] = [];
  for (let i = 1; i < common.length; i++) {
    const p = common[i - 1]!;
    const q = common[i]!;
    ra.push(p.ca > 0 && q.ca > 0 ? Math.log(q.ca / p.ca) : 0);
    rb.push(p.cb > 0 && q.cb > 0 ? Math.log(q.cb / p.cb) : 0);
  }
  return { a: ra, b: rb };
}

/** Correlation of two bar series over their aligned log-returns. */
export function seriesCorrelation(a: Bar[], b: Bar[]): number {
  const { a: ra, b: rb } = alignedLogReturns(a, b);
  return correlation(ra, rb);
}

/**
 * Normalize a close series to a base of 100 at the first bar, for overlay
 * compare charts. Empty in -> empty out; a zero first close yields all-100.
 */
export function normalizeTo100(bars: Bar[]): { ts: number; v: number }[] {
  if (bars.length === 0) return [];
  const base = bars[0]!.c;
  return bars.map((b) => ({ ts: b.ts, v: base !== 0 ? (b.c / base) * 100 : 100 }));
}

export interface BacktestTrade {
  entryIndex: number;
  exitIndex: number;
  entryPrice: number;
  exitPrice: number;
  ret: number;
}

export interface BacktestResult {
  /** Equity curve (starts at 1.0), one entry per bar. */
  equity: number[];
  totalReturn: number;
  trades: BacktestTrade[];
  numTrades: number;
  wins: number;
  /** Fraction of closed trades that were profitable (0 when no trades). */
  winRate: number;
  fast: number;
  slow: number;
}

/**
 * Long/flat SMA-crossover backtest. Go LONG on the bar after fast SMA crosses
 * above slow SMA; go FLAT on the bar after it crosses back below. The position
 * held into a bar is applied to that bar's simple return (no look-ahead).
 * Frictionless.
 */
export function backtestMaCross(bars: Bar[], fast: number, slow: number): BacktestResult {
  const empty: BacktestResult = {
    equity: bars.length ? [1] : [],
    totalReturn: 0,
    trades: [],
    numTrades: 0,
    wins: 0,
    winRate: 0,
    fast,
    slow,
  };
  if (fast <= 0 || slow <= 0 || fast >= slow || bars.length < slow + 1) return empty;

  const closes = bars.map((b) => b.c);
  const fastMa = sma(closes, fast);
  const slowMa = sma(closes, slow);

  const equity: number[] = new Array(bars.length).fill(1);
  const trades: BacktestTrade[] = [];
  let eq = 1;
  let position = 0; // 0 flat, 1 long
  let entryIndex = -1;
  let entryPrice = 0;

  for (let i = 0; i < bars.length; i++) {
    // Apply the position held coming into bar i to this bar's return.
    if (i > 0 && position === 1) {
      const prev = closes[i - 1]!;
      const cur = closes[i]!;
      if (prev !== 0) eq *= cur / prev;
    }
    equity[i] = eq;

    // Decide the signal from THIS bar's completed MAs; it takes effect next bar.
    const f = fastMa[i];
    const s = slowMa[i];
    if (f != null && s != null) {
      const wantLong = f > s;
      if (wantLong && position === 0) {
        position = 1;
        entryIndex = i;
        entryPrice = closes[i]!;
      } else if (!wantLong && position === 1) {
        const exitPrice = closes[i]!;
        trades.push({
          entryIndex,
          exitIndex: i,
          entryPrice,
          exitPrice,
          ret: entryPrice !== 0 ? exitPrice / entryPrice - 1 : 0,
        });
        position = 0;
      }
    }
  }
  // Close any open position at the final bar for accounting.
  if (position === 1 && entryIndex >= 0) {
    const exitPrice = closes[bars.length - 1]!;
    trades.push({
      entryIndex,
      exitIndex: bars.length - 1,
      entryPrice,
      exitPrice,
      ret: entryPrice !== 0 ? exitPrice / entryPrice - 1 : 0,
    });
  }

  const wins = trades.filter((t) => t.ret > 0).length;
  return {
    equity,
    totalReturn: eq - 1,
    trades,
    numTrades: trades.length,
    wins,
    winRate: trades.length ? wins / trades.length : 0,
    fast,
    slow,
  };
}
